server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,309 @@
+                     *     ENTRY_CREATE:operator
+                     * After we get an indication that something has changed, we check the timestamp, file id,
+                     * real path of our desired file.
+                     */
+                    if (Files.exists(settingsDir)) {
+                        try {
+                            Path path = operatorSettingsFile();
+
+                            if (logger.isDebugEnabled()) {
+                                key.pollEvents().stream().forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
+                            }
+
+                            key.pollEvents();
+                            key.reset();
+
+                            enableSettingsWatcher(settingsDir);
+
+                            if (watchedFileChanged(path)) {
+                                processFileSettings(path, false);
+                            }
+                        } catch (Exception e) {
+                            logger.warn("unable to watch or read operator settings file", e);
+                        }
+                    } else {
+                        key.pollEvents();
+                        key.reset();
+                    }
+                }
+            } catch (Exception e) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("encountered exception watching", e);
+                }
+                logger.info("shutting down watcher thread");
+            } finally {
+                watcherThreadLatch.countDown();
+            }
+        }, "elasticsearch[file-settings-watcher]").start();
+    }
+
+    synchronized void stopWatcher() {
+        logger.debug("stopping watcher ...");
+        if (watching()) {
+            try {
+                watchService.close();
+                if (watcherThreadLatch != null) {
+                    watcherThreadLatch.await();
+                }
+            } catch (IOException | InterruptedException e) {
+                logger.info("encountered exception while closing watch service", e);
+            } finally {
+                watchService = null;
+                logger.info("watcher service stopped");
+            }
+        } else {
+            logger.debug("file settings service already stopped");
+        }
+    }
+
+    private void enableSettingsWatcher(Path settingsDir) throws IOException {
+        settingsDir.register(
+            watchService,
+            StandardWatchEventKinds.ENTRY_MODIFY,
+            StandardWatchEventKinds.ENTRY_CREATE,
+            StandardWatchEventKinds.ENTRY_DELETE
+        );
+    }
+
+    void processFileSettings(Path path, boolean onStartup) {
+        logger.info("processing path [{}] for [{}]", path, NAMESPACE);
+        try (
+            XContentParser parser = XContentType.JSON.xContent().createParser(XContentParserConfiguration.EMPTY, Files.newInputStream(path))
```

Comment:
nit: maybe do some static imports and/or set the json xcontent so that the try can be on one line (for readability)

Reply:
Unless I'm mistaken we do not set `StreamReadFeature.AUTO_CLOSE_SOURCE`, so we need to explicitly close the input stream returned from Files::newInputStream, right? Maybe just refactor into the first declaration in the try-with-resources, then pass to createParser. And maybe wrap in a BufferedInputStream ?